File: app/src/main/java/com/designdemo/uaha/DetailActivity.java
package com.designdemo.uaha;

import android.annotation.TargetApi;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.View;
import android.webkit.WebView;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;
import com.support.android.designlibdemo.R;

import java.io.IOException;

public class DetailActivity extends AppCompatActivity {

    public static final String EXTRA_NAME = "os_name";
    public static final String EXTRA_INDEX = "os_index";

    private WebView regsWv;
    private FloatingActionButton fab;
    private CollapsingToolbarLayout collapsingToolbar;

    private OkHttpClient okclient;
    private int osVersion;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail);

        Intent intent = getIntent();
        final String androidName = intent.getStringExtra(EXTRA_NAME);
        osVersion = VersionData.getOsNum(androidName);

        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        collapsingToolbar = (CollapsingToolbarLayout) findViewById(R.id.collapsing_toolbar);
        collapsingToolbar.setTitle(VersionData.getProductName(osVersion));

        okclient = new OkHttpClient();
        loadBackdrop();

        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            postponeEnterTransition();
        }

        setupViews();
        setupFab();
        setupPalette();
        new WikiPullTask().execute();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            if (hasFocus) {
                startPostponedEnterTransition();
            }
        }
    }

    private void loadBackdrop() {
        final ImageView imageView = (ImageView) findViewById(R.id.backdrop);
        Glide.with(this).load(VersionData.getOsDrawable(osVersion)).centerCrop().into(imageView);
    }

    private void setupViews() {
        regsWv = (WebView) findViewById(R.id.regulations_webview);
    }


    private void setupFab() {
        fab = (FloatingActionButton) findViewById((R.id.fab_detail));

        // Set initial state based on pref
        setFabIcon();

        // To over-ride the color of the FAB other then the theme color
        //fab.setBackgroundTintList(ColorStateList.valueOf(getResources().getColor(R.color.grey_300)));

        fab.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                PrefsUtil.toggleFavorite(getApplicationContext(), osVersion);
                setFabIcon();

                //Send the user a message to let them know change was made
                View mainView = findViewById(R.id.main_content);
                Snackbar.make(mainView, R.string.favorite_confirm, Snackbar.LENGTH_LONG)
                        .show(); // Don’t forget to show!
            }
        });
    }

    private void setFabIcon() {
        if (PrefsUtil.isFavorite(getApplicationContext(), osVersion)) {
            fab.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_favorite_on));
        } else {
            fab.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_favorite_off));
        }
    }

    private void setupPalette() {
        Bitmap bm = BitmapFactory.decodeResource(getResources(), VersionData.getOsDrawable(osVersion));

        Palette.PaletteAsyncListener listener = new Palette.PaletteAsyncListener() {
            public void onGenerated(Palette palette) {
                Log.d("Palette", "Palette has been generated");
                TextView pal_vib = (TextView) findViewById(R.id.palette_vibrant);
                TextView perc1left = (TextView) findViewById(R.id.perc1_left);
                perc1left.setBackgroundColor(palette.getVibrantColor(0x000000));
                pal_vib.setBackgroundColor(palette.getVibrantColor(0x000000));

                TextView pal_vib_dark = (TextView) findViewById(R.id.palette_vibrant_dark);
                TextView perc1mid = (TextView) findViewById(R.id.perc1_middle);
                perc1mid.setBackgroundColor(palette.getDarkVibrantColor(0x000000));
                pal_vib_dark.setBackgroundColor(palette.getDarkVibrantColor(0x000000));

                TextView pal_vib_light = (TextView) findViewById(R.id.palette_vibrant_light);
                TextView perc1right = (TextView) findViewById(R.id.perc1_right);
                perc1right.setBackgroundColor(palette.getLightVibrantColor(0x000000));
                pal_vib_light.setBackgroundColor(palette.getLightVibrantColor(0x000000));

                TextView pal_muted = (TextView) findViewById(R.id.palette_muted);
                TextView perc2left = (TextView) findViewById(R.id.perc2_left);
                perc2left.setBackgroundColor(palette.getMutedColor(0x000000));
                pal_muted.setBackgroundColor(palette.getMutedColor(0x000000));

                TextView pal_muted_dark = (TextView) findViewById(R.id.palette_muted_dark);
                TextView perc2mid = (TextView) findViewById(R.id.perc2_middle);
                perc2mid.setBackgroundColor(palette.getDarkMutedColor(0x000000));
                pal_muted_dark.setBackgroundColor(palette.getDarkMutedColor(0x000000));

                TextView pal_muted_light = (TextView) findViewById(R.id.palette_muted_light);
                TextView perc2right = (TextView) findViewById(R.id.perc2_right);
                perc2right.setBackgroundColor(palette.getLightMutedColor(0x000000));
                pal_muted_light.setBackgroundColor(palette.getLightMutedColor(0x000000));

                //Noticed the Expanded white doesn't show everywhere, use Palette to fix this
                collapsingToolbar.setExpandedTitleColor(palette.getVibrantColor(0x000000));
            }
        };

        // Start this Async, because it takes some time to generate
        Palette.from(bm).generate(listener);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.toolbar_actions, menu);
        return true;
    }

    private class WikiPullTask extends AsyncTask<String, Void, String> {
        private String phone;

        @Override
        protected String doInBackground(String... params) {
            String respStr = "";

            String req = "https://en.wikipedia.org/w/api.php?action=query&titles=" + VersionData.getWikiQuery(osVersion) + "&prop=revisions&rvprop=content&format=jsonfm";
            Log.d("NET", "WIKIPedia request: " + req);

            Request request = new Request.Builder()
                    .url(req)
                    .build();

            try {
                Response blacklistResp = okclient.newCall(request).execute();
                respStr = blacklistResp.body().string();
                Log.d("NET", "WIKIPedia_RESPONSE: " + respStr);
            } catch (IOException e) {
                e.printStackTrace();
            }

            return respStr;
        }

        @Override
        protected void onPostExecute(String result) {
            regsWv.loadData(result, "text/html", null);
        }

        @Override
        protected void onPreExecute() {
        }

        @Override
        protected void onProgressUpdate(Void... values) {
        }
    }
}

File: app/src/main/java/com/designdemo/uaha/view/detail/DetailActivity.kt
package com.designdemo.uaha.view.detail

import android.annotation.TargetApi
import android.app.SearchManager
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.palette.graphics.Palette
import com.bumptech.glide.Glide
import com.designdemo.uaha.data.model.detail.DetailEntity
import com.designdemo.uaha.data.model.product.ProductEntity
import com.designdemo.uaha.util.UiUtil
import com.designdemo.uaha.view.demo.BottomNavActivity
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.snackbar.Snackbar
import com.support.android.designlibdemo.R
import com.support.android.designlibdemo.R.color.black
import kotlinx.android.synthetic.main.activity_detail.*
import kotlin.random.Random

class DetailActivity : AppCompatActivity() {

    // Details Card Views
    private var iconColor = 0

    private var localProdItem: ProductEntity? = null

    private lateinit var detailViewModel: DetailViewModel

    private var androidName = "unset"

    companion object {
        private const val TAG = "DetailActivity"

        const val EXTRA_APP_NAME = "os_name"
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail)

        if (androidName === "unset") {
            androidName = intent.getStringExtra(EXTRA_APP_NAME)
        }

        detailViewModel = ViewModelProviders.of(this).get(DetailViewModel::class.java)

        if (Intent.ACTION_SEARCH == intent.action) {
            val query = intent.getStringExtra(SearchManager.QUERY)
        }

        spec_layout.visibility = View.GONE
    }

    override fun onResume() {
        super.onResume()

        val intent = intent
        handleIntent(intent)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            postponeEnterTransition()
        }

        setupFab()

        val product = androidName.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        detailViewModel.getProductForName(product[0]).observe(this, Observer { prodItem ->
            prodItem?.let {
                localProdItem = prodItem
                Glide.with(this).load(prodItem.imgId).centerCrop().into(backdrop)
                setFabIcon()
                setupPalette(prodItem.imgId!!)
                setupToolbar()
            }
        })

        detailViewModel.getDetailsForItem(androidName).observe(this, Observer { detailsForItem ->
            detailsForItem?.let {
                setDeviceInfoViews(detailsForItem)
            }
        })

        detailViewModel.getProgressBarVisibility().observe(this, Observer { progVisibility ->
            if (progVisibility) {
                details_progress_layout.visibility = View.VISIBLE
            } else {
                details_progress_layout.visibility = View.GONE
            }
        })

        // Retrieve the phone and WIKI info from net, and put in DB when found
        detailViewModel.getFonoNetInfo(androidName)
//        detailViewModel.getWikiNetInfo("Android Oreo")
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            Log.d(TAG, "The search query is happenning ${intent.getStringExtra(SearchManager.QUERY)}")
            val query = intent.getStringExtra(SearchManager.QUERY)
        }
    }

    // TODO, setup this better
    private fun setupToolbar() {
        setSupportActionBar(detail_toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        bottom_appbar.replaceMenu(R.menu.detail_actions)
        bottom_appbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.menu_shuffle -> {
                    // Get a random Android product, and then refresh the UI
                    // TODO need to move to VM, because it is broken here
                    clearDeviceInfoViews()
                    val deviceStrings = resources.getStringArray(R.array.device_names)
                    val randInt = Random.nextInt(deviceStrings.size - 1)
                    // TODO Send this value to VM, and the force a refresh of views with data from this item
                    detailViewModel.updateProductFromRefresh(deviceStrings[randInt])
                    this.androidName = deviceStrings[randInt]
//                    onResume()
                }
                R.id.menu_help -> {
                    val bottomNavIntent = Intent(applicationContext, BottomNavActivity::class.java)
                    startActivity(bottomNavIntent)
                }
            }
            false
        }

        collapsing_toolbar.title = localProdItem!!.title
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && hasFocus) {
            startPostponedEnterTransition()
        }
    }

    private fun setupFab() {
        fab_detail.setOnClickListener {
            // Toggle the value to be opposite of what it is, to set
            var toSetFav = 1
            if (localProdItem?.isFav == 1) {
                toSetFav = 0
            }

            localProdItem?.apply {
                isFav = toSetFav
            }?.let { item ->
                detailViewModel.insertFav(item)
            }

            // Notify the User with a snackbar
            // Need to set the FAB as the anchor view, so this appears higher then the BottomAppBar per Material Spec
            Snackbar.make(detail_main_linear, R.string.favorite_confirm, Snackbar.LENGTH_LONG)
                    .setAnchorView(fab_detail)
                    .show()
        }
    }

    private fun setFabIcon() {
        if (localProdItem!!.isFav == 1) {
            fab_detail.isSelected = true
            fab_detail.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.vct_fav_white_sel))
        } else {
            fab_detail.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.vct_fav_white_unsel))
        }
    }

    @Suppress("UnnecessaryParentheses")
    private fun setupPalette(imgId: Int) {
        val bm = BitmapFactory.decodeResource(resources, imgId)
        try {
            Palette.from(bm).generate { palette ->
                Log.d(TAG, "Palette has been generated")
                perc1_left.setBackgroundColor(palette!!.getVibrantColor(0x000000))
                palette_vibrant.setBackgroundColor(palette.getVibrantColor(0x000000))

                perc1_middle.setBackgroundColor(palette.getDarkVibrantColor(0x000000))
                palette_vibrant_dark.setBackgroundColor(palette.getDarkVibrantColor(0x000000))

                perc1_right.setBackgroundColor(palette.getLightVibrantColor(0x000000))
                palette_vibrant_light.setBackgroundColor(palette.getLightVibrantColor(0x000000))

                perc2_left.setBackgroundColor(palette.getMutedColor(0x000000))
                palette_muted.setBackgroundColor(palette.getMutedColor(0x000000))

                perc2_middle.setBackgroundColor(palette.getDarkMutedColor(0x000000))
                palette_muted_dark.setBackgroundColor(palette.getDarkMutedColor(0x000000))

                perc2_right.setBackgroundColor(palette.getLightMutedColor(0x000000))
                palette_muted_light.setBackgroundColor(palette.getLightMutedColor(0x000000))

                // Noticed the Expanded white doesn't show everywhere, use Palette to fix this
                collapsing_toolbar?.setExpandedTitleColor(palette.getVibrantColor(0x000000))

                iconColor = if (palette.getVibrantColor(0x000000) == 0x000000)
                    palette.getMutedColor(0x000000)
                else
                    palette.getVibrantColor(0x000000)

                // The back button should also have a better color applied to ensure it is visible,
                val res = resources.getIdentifier("abc_ic_ab_back_material", "drawable", packageName)
                val upArrow = getColorizedDrawable(res)
                val whiteArrow = getColorizedDrawable(res, R.color.white_grey)

                // This applies the custom color to the home button, when expanded, and white when collapsed
                appbar?.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener {
                    appBarLayout: AppBarLayout, offset: Int ->
                    val isCollapsed = (offset == (-1 * appBarLayout.totalScrollRange))
                    if (!isCollapsed)
                        supportActionBar?.setHomeAsUpIndicator(upArrow)
                    else
                        supportActionBar?.setHomeAsUpIndicator(whiteArrow)
                })
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error with image! + $e", e)
        }
    }

    private fun setDeviceInfoViews(detailItem: DetailEntity) {
        details_device_name.text = detailItem.DeviceName
        details_features.text = detailItem.features
        details_features_c.text = detailItem.features_c

        if (detailItem.multitouch != null)
            setupSpecItem(R.drawable.vct_multitouch, R.string.spec_multitouch, detailItem.multitouch, spec_multitouch)
        else
            setupSpecItem(R.drawable.vct_multitouch, R.string.spec_multitouch, "NA", spec_multitouch)

        val cpuInfo = detailItem.chipset + "\n" + detailItem.cpu
        setupSpecItem(R.drawable.vct_cpu, R.string.spec_cpu, cpuInfo, spec_cpu)

        if (detailItem.announced != null)
            setupSpecItem(R.drawable.vct_date, R.string.spec_release_date, detailItem.announced, spec_releasedate)

        if (detailItem.size != null)
            setupSpecItem(R.drawable.vct_size, R.string.spec_size, detailItem.size, spec_size)

        if (detailItem.weight != null)
            setupSpecItem(R.drawable.vct_weight, R.string.spec_weight, detailItem.weight, spec_weight)

        if (detailItem.dimensions != null)
            setupSpecItem(R.drawable.vct_dimen, R.string.spec_dimen, detailItem.dimensions, spec_dimen)

        if (detailItem.internal != null)
            setupSpecItem(R.drawable.vct_memory, R.string.spec_memory, detailItem.internal, spec_memory)

        if (detailItem.primary_ != null)
            setupSpecItem(R.drawable.vct_camera, R.string.spec_camera, detailItem.primary_, spec_camera)

        if (detailItem.video != null)
            setupSpecItem(R.drawable.vct_video, R.string.spec_video, detailItem.video, spec_video)

        if (detailItem.type != null)
            setupSpecItem(R.drawable.vct_display, R.string.spec_display, detailItem.type, spec_display)

        if (detailItem.resolution != null)
            setupSpecItem(R.drawable.vct_resolution, R.string.spec_resolution, detailItem.resolution, spec_resolution)

        spec_layout.visibility = View.VISIBLE
    }

    private fun clearDeviceInfoViews() {
        details_device_name.text = getString(R.string.getting_information)
        details_device_name.setTextColor(resources.getColor(black))
        details_features.text = ""
        details_features_c.text = ""

        setupSpecItem(R.drawable.vct_multitouch, R.string.spec_multitouch, "", spec_multitouch)
        setupSpecItem(R.drawable.vct_cpu, R.string.spec_cpu, "", spec_cpu)
        setupSpecItem(R.drawable.vct_date, R.string.spec_release_date, "", spec_releasedate)
        setupSpecItem(R.drawable.vct_size, R.string.spec_size, "", spec_size)
        setupSpecItem(R.drawable.vct_weight, R.string.spec_weight, "", spec_weight)
        setupSpecItem(R.drawable.vct_dimen, R.string.spec_dimen, "", spec_dimen)
        setupSpecItem(R.drawable.vct_memory, R.string.spec_memory, "", spec_memory)
        setupSpecItem(R.drawable.vct_camera, R.string.spec_camera, "", spec_camera)
        setupSpecItem(R.drawable.vct_video, R.string.spec_video, "", spec_video)
        setupSpecItem(R.drawable.vct_display, R.string.spec_display, "", spec_display)
        setupSpecItem(R.drawable.vct_resolution, R.string.spec_resolution, "", spec_resolution)
        spec_layout.visibility = View.VISIBLE
    }

    private fun setupSpecItem(@DrawableRes drawable: Int, @StringRes title: Int, info: String, view: TextView) {
        val specDrawable = getColorizedDrawable(drawable)
        view.setCompoundDrawablesWithIntrinsicBounds(specDrawable, null, null, null)
        view.text = UiUtil.applyBoldFirstWord(getString(title), info)
    }

    private fun getColorizedDrawable(@DrawableRes res: Int): Drawable? {
        val drawable = ContextCompat.getDrawable(this, res)
        drawable?.setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP)
        return drawable
    }

    private fun getColorizedDrawable(@DrawableRes res: Int, color: Int): Drawable? {
        val drawable = ContextCompat.getDrawable(this, res)
        drawable?.setColorFilter(resources.getColor(color), PorterDuff.Mode.SRC_ATOP)
        return drawable
    }
}
